test/unit/bigip/test_bigip_software_facts.py

```diff
@@ -0,0 +1,389 @@
+            server='localhost',
+            password='password',
+            user='admin',
+            include='hotfix'
+        ))
+
+        client = AnsibleF5Client(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            f5_product_name=self.spec.f5_product_name
+        )
+
+        mm = FactManagerBase(client)
+        mm.exit_json = Mock(return_value=False)
+        results = mm.display_facts()
+
+        assert results['changed'] is True
+        assert 'images' not in results.keys()
+        assert 'volumes' not in results.keys()
+        assert len(results['hotfixes']) == 5
+
+    def test_get_volumes_no_filter(self, *args):
+        set_module_args(dict(
+            server='localhost',
+            password='password',
+            user='admin',
+            include='volume'
+        ))
+
+        client = AnsibleF5Client(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            f5_product_name=self.spec.f5_product_name
+        )
+
+        mm = FactManagerBase(client)
+        mm.exit_json = Mock(return_value=False)
+        results = mm.display_facts()
+
+        assert results['changed'] is True
+        assert 'images' not in results.keys()
+        assert 'hotfixes' not in results.keys()
+        assert len(results['volumes']) == 4
+
+    def test_get_image_and_hotfixes_no_filter(self, *args):
+        set_module_args(dict(
+            server='localhost',
+            password='password',
+            user='admin',
+            include=['image', 'hotfix']
+        ))
+
+        client = AnsibleF5Client(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            f5_product_name=self.spec.f5_product_name
+        )
+
+        mm = FactManagerBase(client)
+        mm.exit_json = Mock(return_value=False)
+        results = mm.display_facts()
+
+        assert results['changed'] is True
+        assert len(results['images']) == 8
+        assert len(results['hotfixes']) == 5
+        assert 'volumes' not in results.keys()
+
+    def test_get_all_facts_filter_on_version(self, *args):
+        set_module_args(dict(
+            server='localhost',
+            password='password',
+            user='admin',
+            include='all',
+            filter='version:12.1.1'
+        ))
+
+        client = AnsibleF5Client(
+            argument_spec=self.spec.argument_spec,
+            supports_check_mode=self.spec.supports_check_mode,
+            f5_product_name=self.spec.f5_product_name
+        )
+
+        mm = FactManagerBase(client)
+        mm.exit_json = Mock(return_value=False)
+        results = mm.display_facts()
+
+        assert results['changed'] is True
+        assert len(results['hotfixes']) == 0
+        assert results['images'] == [{'product': 'BIG-IP', 'name': 'BIGIP-12.1.1.0.0.184.iso',
```

Comment:
it's a bad idea to assert the structure of a dictionary because dicts are by nature, unstructured. Instead, assert the following

* len of the list == 1
* that the product key exists in [0]
* that the name key exists in [0]
* that the value of [0]['product'] is 'BIG-IP'
* that the value of [0]['product'] is 'BIGIP-12.1.1.0.0.184.iso'

Do this change for _all_ dict assertions